File: stat_react/server/mstockScripMaster.mjs
/**
 * Build NIFTY option chain from OpenAPIScripMaster + batch LTP quote (Type B).
 * Fallback when GetOptionChain is blocked or unavailable.
 */

import { httpsGet } from './mstockHttps.mjs';
import { quoteHeaders, fetchNseTokenLtps } from './niftyQuote.mjs';
import {
  atmStrike,
  nearestNiftyWeeklyExpiry,
  OPTION_CHAIN_STRIKE_COUNT,
  strikesFromSpot,
} from './analysis.mjs';
import {
  clearMstockIpBlock,
  markMstockIpBlocked,
  throwIfMstockIpError,
} from './mstockApiGuard.mjs';

const SCRIP_PATH = '/openapi/typeb/instruments/OpenAPIScripMaster';

const MONTHS = {
  JAN: 1,
  FEB: 2,
  MAR: 3,
  APR: 4,
  MAY: 5,
  JUN: 6,
  JUL: 7,
  AUG: 8,
  SEP: 9,
  OCT: 10,
  NOV: 11,
  DEC: 12,
};

/** @param {string} expiryIso YYYY-MM-DD */
function expiryIsoToDdMmmYy(iso) {
  const [y, m, d] = iso.split('-').map(Number);
  if (!y || !m || !d) return '';
  const mon = Object.keys(MONTHS).find((k) => MONTHS[k] === m) ?? 'JAN';
  return `${String(d).padStart(2, '0')}${mon}${String(y).slice(-2)}`;
}

/** NSE NIFTY index options lot (override via MSTOCK_NIFTY_LOT_SIZE). */
export function defaultNiftyLotSize() {
  const n = Number(process.env.MSTOCK_NIFTY_LOT_SIZE || 75);
  return Number.isFinite(n) && n > 0 ? Math.round(n) : 75;
}

/** @param {Record<string, unknown>} row */
export function parseInstrumentLotSize(row) {
  const raw =
    row.lotsize ?? row.lot_size ?? row.LotSize ?? row.lotSize ?? row['lot size'] ?? '';
  const n = Number(String(raw).replace(/,/g, ''));
  if (Number.isFinite(n) && n >= 1 && n <= 1000) return Math.round(n);
  return defaultNiftyLotSize();
}

/** Broker tradingsymbol for place-order (full option name, not underlying only). */
export function resolveTradingsymbol(row) {
  const candidates = [
    row.tradingsymbol,
    row.trading_symbol,
    row.TradingSymbol,
    row.name,
    row.symbol,
  ]
    .map((s) => String(s ?? '').trim())
    .filter(Boolean);
  const optionLeg = candidates.find((c) => /(CE|PE)$/i.test(c));
  return optionLeg || candidates[0] || '';
}

/** @param {Record<string, unknown>} row */
function normalizeInstrument(row) {
  if (!row || typeof row !== 'object') return row;
  return {
    ...row,
    lotsize: String(parseInstrumentLotSize(row)),
    tradingsymbol: resolveTradingsymbol(row),
  };
}

/** @param {unknown} raw */
function parseInstruments(raw) {
  if (!raw || typeof raw !== 'string') return [];
  const t = raw.trim();
  if (t.startsWith('[')) {
    try {
      const arr = JSON.parse(t);
      return Array.isArray(arr) ? arr.map((row) => normalizeInstrument(row)) : [];
    } catch {
      return [];
    }
  }
  const rows = [];
  const lines = t.split(/\r?\n/).filter((l) => l.trim());
  if (lines.length < 2) return rows;
  const header = lines[0].toLowerCase();
  const sep = header.includes('\t') ? '\t' : ',';
  const cols = lines[0].split(sep).map((c) => c.trim().toLowerCase());
  const idx = (...names) => {
    for (const name of names) {
      const j = cols.indexOf(name.toLowerCase());
      if (j >= 0) return j;
    }
    return -1;
  };
  const get = (parts, ...names) => {
    const j = idx(...names);
    return j >= 0 ? parts[j]?.trim() : '';
  };
  for (let i = 1; i < lines.length; i++) {
    const parts = lines[i].split(sep);
    rows.push(
      normalizeInstrument({
        token: get(parts, 'token'),
        symbol: get(parts, 'symbol'),
        name: get(parts, 'name'),
        expiry: get(parts, 'expiry'),
        strike: get(parts, 'strike'),
        lotsize: get(parts, 'lotsize', 'lot_size'),
        tradingsymbol: get(parts, 'tradingsymbol', 'trading_symbol', 'tradingsym'),
        instrumenttype: get(parts, 'instrumenttype', 'instrument_type'),
        exch_seg: get(parts, 'exch_seg', 'exchange', 'segment'),
      })
    );
  }
  return rows;
}

function isNiftyOption(row) {
  const sym = String(row.symbol ?? '').toUpperCase();
  const name = String(row.name ?? '').toUpperCase();
  const type = String(row.instrumenttype ?? '').toUpperCase();
  const seg = String(row.exch_seg ?? '').toUpperCase();
  if (!sym.includes('NIFTY') && !name.includes('NIFTY')) return false;
  if (type && !type.includes('OPT')) return false;
  if (seg && !/NFO|NSE|FO|BFO/.test(seg)) return false;
  return /CE|PE/.test(name) || /CE|PE/.test(sym);
}

function parseStrike(row) {
  let strike = Number(row.strike);
  if (Number.isFinite(strike) && strike > 0) {
    if (strike > 50_000) strike = Math.round(strike / 100);
    return strike;
  }
  const name = String(row.name ?? row.symbol ?? '').toUpperCase();
  const m = name.match(/(\d{4,5})(CE|PE)$/);
  if (m) return Number(m[1]);
  return null;
}

function isCe(row) {
  const name = String(row.name ?? row.symbol ?? '').toUpperCase();
  return name.endsWith('CE');
}

function isPe(row) {
  const name = String(row.name ?? row.symbol ?? '').toUpperCase();
  return name.endsWith('PE');
}

/** @param {string} expiryField @param {string} expiryIso */
function expiryMatches(expiryField, expiryIso) {
  const tag = expiryIsoToDdMmmYy(expiryIso);
  if (!tag) return true;
  const e = String(expiryField ?? '').toUpperCase().replace(/\s/g, '');
  const name = String(expiryField ?? '').toUpperCase();
  if (e.includes(tag)) return true;
  const [y, m, d] = expiryIso.split('-');
  const mon = Object.keys(MONTHS).find((k) => MONTHS[k] === Number(m));
  if (mon && e.includes(`${d}${mon}`)) return true;
  return false;
}

/**
 * @param {string} apiKey
 * @param {string} jwt
 * @param {number} spot
 */
export async function fetchNiftyChainFromScripMaster(apiKey, jwt, spot) {
  const headers = quoteHeaders(apiKey, jwt);
  const { statusCode, text } = await httpsGet(SCRIP_PATH, headers);
  if (statusCode === 200) clearMstockIpBlock();
  throwIfMstockIpError(statusCode, text);
  if (statusCode !== 200) {
    throw new Error(`scrip master HTTP ${statusCode}: ${text.slice(0, 160)}`);
  }

  const instruments = parseInstruments(text);
  const expiryIso = nearestNiftyWeeklyExpiry();
  const atm = atmStrike(spot, 50);
  const minStrike = strikesFromSpot(spot, 50, 1)[0] ?? atm;
  const byStrike = new Map();

  for (const row of instruments) {
    if (!isNiftyOption(row)) continue;
    if (!expiryMatches(row.expiry || row.name, expiryIso)) continue;
    const strike = parseStrike(row);
    if (strike == null) continue;
    if (strike < minStrike) continue;

    const cur = byStrike.get(strike) ?? {};
    const token = String(row.token ?? '').trim();
    if (!token) continue;
    if (isCe(row)) cur.ceToken = token;
    if (isPe(row)) cur.peToken = token;
    byStrike.set(strike, cur);
  }

  const strikes = [...byStrike.keys()]
    .filter((s) => {
      const l = byStrike.get(s);
      return l?.ceToken || l?.peToken;
    })
    .sort((a, b) => a - b);

  const nearest = strikes.filter((s) => s >= minStrike).slice(0, OPTION_CHAIN_STRIKE_COUNT);

  if (!nearest.length) {
    throw new Error('No NIFTY option legs in scrip master for this expiry');
  }

  const tokens = [];
  for (const strike of nearest) {
    const legs = byStrike.get(strike);
    if (legs?.ceToken) tokens.push(legs.ceToken);
    if (legs?.peToken) tokens.push(legs.peToken);
  }

  const ltps = await fetchNseTokenLtps(apiKey, jwt, tokens);
  if (ltps.size > 0) clearMstockIpBlock();

  const chain = nearest.map((strike) => {
    const legs = byStrike.get(strike) ?? {};
    const ceLtp = legs.ceToken ? ltps.get(legs.ceToken) : null;
    const peLtp = legs.peToken ? ltps.get(legs.peToken) : null;
    return {
      strike,
      ce: {
        ltp: ceLtp != null ? Math.round(ceLtp * 100) / 100 : 0,
        oiChangePct: 0,
        volume: 0,
      },
      pe: {
        ltp: peLtp != null ? Math.round(peLtp * 100) / 100 : 0,
        oiChangePct: 0,
        volume: 0,
      },
    };
  });

  const withLtp = chain.filter((r) => r.ce.ltp > 0 || r.pe.ltp > 0).length;
  if (!withLtp) {
    throw new Error('Scrip master found legs but LTP quote returned no prices');
  }

  return { chain, atm, expiryIso, source: 'mstock' };
}

let scripInstrumentsCache = { at: 0, rows: [] };

/** @param {string} apiKey @param {string} jwt */
async function loadScripInstruments(apiKey, jwt) {
  const maxAge = Number(process.env.MSTOCK_SCRIP_CACHE_MS || 300_000);
  if (scripInstrumentsCache.rows.length && Date.now() - scripInstrumentsCache.at < maxAge) {
    return scripInstrumentsCache.rows;
  }
  const headers = quoteHeaders(apiKey, jwt);
  const { statusCode, text } = await httpsGet(SCRIP_PATH, headers);
  if (statusCode === 200) clearMstockIpBlock();
  throwIfMstockIpError(statusCode, text);
  if (statusCode !== 200) {
    throw new Error(`scrip master HTTP ${statusCode}: ${text.slice(0, 160)}`);
  }
  const rows = parseInstruments(text);
  scripInstrumentsCache = { at: Date.now(), rows };
  return rows;
}

/**
 * Resolve NIFTY weekly option leg for place-order (tradingsymbol + token).
 * @param {string} apiKey
 * @param {string} jwt
 * @param {number} strike
 * @param {'CE'|'PE'} optionType
 */
export async function findNiftyOptionInstrument(apiKey, jwt, strike, optionType) {
  const instruments = await loadScripInstruments(apiKey, jwt);
  const expiryIso = nearestNiftyWeeklyExpiry();
  const wantCe = optionType === 'CE';
  for (const row of instruments) {
    if (!isNiftyOption(row)) continue;
    if (!expiryMatches(row.expiry || row.name, expiryIso)) continue;
    const s = parseStrike(row);
    if (s !== strike) continue;
    if (wantCe && !isCe(row)) continue;
    if (!wantCe && !isPe(row)) continue;
    const seg = String(row.exch_seg ?? 'NFO').toUpperCase();
    const exchange = /BFO/.test(seg) ? 'BFO' : 'NFO';
    const tradingsymbol = resolveTradingsymbol(row);
    const symboltoken = String(row.token ?? '').trim();
    const lotsize = parseInstrumentLotSize(row);
    if (!tradingsymbol || !symboltoken) continue;
    if (!/(CE|PE)$/i.test(tradingsymbol)) continue;
    return { tradingsymbol, symboltoken, exchange, lotsize };
  }
  return null;
}

File: lib/strategy_research/data/sources/remote/mstock_scrip_master.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../../auth/mstock_jwt_manager.dart';
import '../../constants/strategy_research_api_keys.dart';
import 'mstock_api_helpers.dart';

const _months = {
  'JAN': 1,
  'FEB': 2,
  'MAR': 3,
  'APR': 4,
  'MAY': 5,
  'JUN': 6,
  'JUL': 7,
  'AUG': 8,
  'SEP': 9,
  'OCT': 10,
  'NOV': 11,
  'DEC': 12,
};

class NiftyOptionLeg {
  const NiftyOptionLeg({
    required this.tradingsymbol,
    required this.symboltoken,
    required this.exchange,
    required this.lotsize,
  });

  final String tradingsymbol;
  final String symboltoken;
  final String exchange;
  final int lotsize;
}

int defaultNiftyLotSize() => 75;

int parseInstrumentLotSize(Map<String, dynamic> row) {
  for (final key in ['lotsize', 'lot_size', 'LotSize', 'lotSize']) {
    final raw = row[key];
    if (raw == null) continue;
    final n = int.tryParse(raw.toString().replaceAll(',', ''));
    if (n != null && n >= 1 && n <= 1000) return n;
  }
  return defaultNiftyLotSize();
}

String resolveTradingsymbol(Map<String, dynamic> row) {
  final candidates = [
    row['tradingsymbol'],
    row['trading_symbol'],
    row['TradingSymbol'],
    row['name'],
    row['symbol'],
  ]
      .map((s) => s?.toString().trim() ?? '')
      .where((s) => s.isNotEmpty)
      .toList();
  for (final c in candidates) {
    if (RegExp(r'(CE|PE)$', caseSensitive: false).hasMatch(c)) return c;
  }
  return candidates.isNotEmpty ? candidates.first : '';
}

List<Map<String, dynamic>> parseScripInstruments(String raw) {
  final t = raw.trim();
  if (t.startsWith('[')) {
    try {
      final arr = jsonDecode(t) as List<dynamic>;
      return arr
          .whereType<Map>()
          .map((e) => Map<String, dynamic>.from(e))
          .toList();
    } catch (_) {
      return [];
    }
  }
  return [];
}

String nearestNiftyWeeklyExpiryIso([DateTime? now]) {
  final ist = (now ?? DateTime.now()).toUtc().add(const Duration(hours: 5, minutes: 30));
  var candidate = DateTime.utc(ist.year, ist.month, ist.day);
  while (candidate.weekday != DateTime.tuesday) {
    candidate = candidate.add(const Duration(days: 1));
  }
  if (ist.weekday == DateTime.tuesday &&
      (ist.hour > 15 || (ist.hour == 15 && ist.minute >= 30))) {
    candidate = candidate.add(const Duration(days: 7));
  }
  return '${candidate.year.toString().padLeft(4, '0')}-'
      '${candidate.month.toString().padLeft(2, '0')}-'
      '${candidate.day.toString().padLeft(2, '0')}';
}

bool _isNiftyOption(Map<String, dynamic> row) {
  final sym = row['symbol']?.toString().toUpperCase() ?? '';
  final name = row['name']?.toString().toUpperCase() ?? '';
  final type = row['instrumenttype']?.toString().toUpperCase() ?? '';
  final seg = row['exch_seg']?.toString().toUpperCase() ?? '';
  if (!sym.contains('NIFTY') && !name.contains('NIFTY')) return false;
  if (type.isNotEmpty && !type.contains('OPT')) return false;
  if (seg.isNotEmpty && !RegExp(r'NFO|NSE|FO|BFO').hasMatch(seg)) return false;
  return RegExp(r'CE|PE').hasMatch(name) || RegExp(r'CE|PE').hasMatch(sym);
}

int? _parseStrike(Map<String, dynamic> row) {
  var strike = num.tryParse(row['strike']?.toString() ?? '');
  if (strike != null && strike > 0) {
    if (strike > 50000) strike = (strike / 100).round();
    return strike.round();
  }
  final name = (row['name'] ?? row['symbol'] ?? '').toString().toUpperCase();
  final m = RegExp(r'(\d{4,5})(CE|PE)$').firstMatch(name);
  if (m != null) return int.tryParse(m.group(1)!);
  return null;
}

bool _isCe(Map<String, dynamic> row) {
  final name = (row['name'] ?? row['symbol'] ?? '').toString().toUpperCase();
  return name.endsWith('CE');
}

bool _isPe(Map<String, dynamic> row) {
  final name = (row['name'] ?? row['symbol'] ?? '').toString().toUpperCase();
  return name.endsWith('PE');
}

String _expiryIsoToDdMmmYy(String iso) {
  final parts = iso.split('-');
  if (parts.length != 3) return '';
  final y = int.tryParse(parts[0]);
  final m = int.tryParse(parts[1]);
  final d = int.tryParse(parts[2]);
  if (y == null || m == null || d == null) return '';
  final mon = _months.entries
      .firstWhere((e) => e.value == m, orElse: () => const MapEntry('JAN', 1))
      .key;
  final yy = (y % 100).toString().padLeft(2, '0');
  return '${d.toString().padLeft(2, '0')}$mon$yy';
}

bool _expiryMatches(String? expiryField, String expiryIso) {
  final tag = _expiryIsoToDdMmmYy(expiryIso);
  if (tag.isEmpty) return true;
  final e = (expiryField ?? '').toUpperCase().replaceAll(' ', '');
  if (e.contains(tag)) return true;
  final parts = expiryIso.split('-');
  if (parts.length == 3) {
    final monthNum = int.tryParse(parts[1]);
    final mon = _months.entries
        .firstWhere(
          (entry) => entry.value == monthNum,
          orElse: () => const MapEntry('JAN', 1),
        )
        .key;
    if (e.contains('${parts[2]}$mon')) return true;
  }
  return false;
}

/// Resolve NIFTY weekly option leg for place-order (mirrors stat_react).
class MstockScripMaster {
  MstockScripMaster({http.Client? client}) : _client = client ?? http.Client();

  final http.Client _client;
  List<Map<String, dynamic>> _cache = [];
  int _cacheAt = 0;

  Future<List<Map<String, dynamic>>> loadInstruments() async {
    const maxAgeMs = 300000;
    if (_cache.isNotEmpty &&
        DateTime.now().millisecondsSinceEpoch - _cacheAt < maxAgeMs) {
      return _cache;
    }
    final apiKey = StrategyResearchApiKeys.apiKey;
    if (apiKey.isEmpty) return [];
    final uri = Uri.parse(
      '${StrategyResearchApiKeys.baseUrl}${StrategyResearchApiKeys.scripMasterPath}',
    );
    final headers = mstockQuoteHeaders(apiKey, StrategyResearchApiKeys.jwtToken);
    var response = await _client.get(uri, headers: headers).timeout(
          const Duration(seconds: 30),
        );
    if (response.statusCode == 401) {
      final refreshed = await MstockJwtManager.instance.refreshFromTotp();
      if (refreshed) {
        response = await _client
            .get(
              uri,
              headers: mstockQuoteHeaders(
                apiKey,
                StrategyResearchApiKeys.jwtToken,
              ),
            )
            .timeout(const Duration(seconds: 30));
      }
    }
    if (response.statusCode != 200) return [];
    _cache = parseScripInstruments(response.body);
    _cacheAt = DateTime.now().millisecondsSinceEpoch;
    return _cache;
  }

  Future<NiftyOptionLeg?> findNiftyOptionInstrument(
    int strike,
    String optionType,
  ) async {
    final instruments = await loadInstruments();
    final expiryIso = nearestNiftyWeeklyExpiryIso();
    final wantCe = optionType.toUpperCase() == 'CE';
    for (final row in instruments) {
      if (!_isNiftyOption(row)) continue;
      if (!_expiryMatches(
        row['expiry']?.toString() ?? row['name']?.toString(),
        expiryIso,
      )) {
        continue;
      }
      if (_parseStrike(row) != strike) continue;
      if (wantCe && !_isCe(row)) continue;
      if (!wantCe && !_isPe(row)) continue;
      final seg = row['exch_seg']?.toString().toUpperCase() ?? 'NFO';
      final exchange = seg.contains('BFO') ? 'BFO' : 'NFO';
      final tradingsymbol = resolveTradingsymbol(row);
      final symboltoken = row['token']?.toString().trim() ?? '';
      if (tradingsymbol.isEmpty || symboltoken.isEmpty) continue;
      if (!RegExp(r'(CE|PE)$', caseSensitive: false).hasMatch(tradingsymbol)) {
        continue;
      }
      return NiftyOptionLeg(
        tradingsymbol: tradingsymbol,
        symboltoken: symboltoken,
        exchange: exchange,
        lotsize: parseInstrumentLotSize(row),
      );
    }
    return null;
  }
}

/// Parse strike + CE/PE from broker tradingsymbol (e.g. NIFTY2560525000CE).
({int strike, String optionType})? parseNiftyOptionSymbol(String tradingsymbol) {
  final t = tradingsymbol.toUpperCase();
  final m = RegExp(r'NIFTY.*?(\d{4,5})(CE|PE)$').firstMatch(t);
  if (m == null) return null;
  final strike = int.tryParse(m.group(1)!);
  if (strike == null) return null;
  return (strike: strike, optionType: m.group(2)!);
}
